### PR TITLE
os/bluestore: fix deferred writes vs collection split race

### PR DESCRIPTION
**Pull request by liewegas**
Fixes: http://tracker.ceph.com/issues/19379

The first patch was prelim work for an alternate solution, and not
needed here, but I think it's a nice change anyway.

### CODE REVIEW

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -1603,6 +1603,12 @@ class BlueStore : public ObjectStore,
 	qcond.wait(l);
     }
 
+    void drain_preceding(TransContext *txc) {
+      std::unique_lock<std::mutex> l(qlock);
+      while (!q.empty() && &q.back() != txc)
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: wondering if that should be q.front() != txc

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -7561,6 +7566,28 @@ bool BlueStore::_osr_reap_done(OpSequencer *osr)
   return empty;
 }
 
+void BlueStore::_osr_drain_preceding(TransContext *txc)
+{
+  OpSequencer *osr = txc->osr.get();
+  dout(10) << __func__ << " " << txc << " osr " << osr << dendl;
+  deferred_aggressive = true; // FIXME: maybe osr-local aggressive flag?
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Indeed making deferred_aggressive osr-local makes sense IMO

**Comment by liewegas** [comment]
yes!  fixed

**Comment by liewegas** [comment]
fixed build

**Pull request merged**